overseer/src/lib.rs

```diff
@@ -0,0 +1,623 @@
+///
+/// It is generic over some `M` that is intended to be a message type
+/// being used by the subsystems running on the `Overseer`. Most likely
+/// this type will be one large `enum` covering all possible messages in
+/// the system.
+/// It is also generic over `I` that is entended to be a type identifying
+/// different subsystems, again most likely this is a one large `enum`
+/// covering all possible subsystem kinds.
+enum OverseerMessage<M: Debug, I> {
+	/// This is a message generated by a `Subsystem`.
+	/// Wraps the messge itself and has an optional `to` of
+	/// someone who can receive this message.
+	///
+	/// If that `to` is present the message will be targetedly sent to the intended
+	/// receiver. The most obvious use case of this is communicating with children.
+	SubsystemMessage {
+		to: Option<I>,
+		msg: M,
+	},
+	/// A message that wraps something the `Subsystem` is desiring to
+	/// spawn on the overseer and a `oneshot::Sender` to signal the result
+	/// of the spawn.
+	SpawnChild {
+		s: (I, Box<dyn Subsystem<M, I> + Send>),
+		res: oneshot::Sender<Result<I, SubsystemError>>,
+	},
+}
+
+impl<M: Debug, I: Debug> Debug for OverseerMessage<M, I> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		match self {
+			OverseerMessage::SubsystemMessage { to, msg } => {
+				write!(f, "OverseerMessage::SubsystemMessage{{ to: {:?}, msg: {:?} }}", to, msg)
+			}
+			OverseerMessage::SpawnChild { .. } => write!(f, "OverseerMessage::Spawn(..)")
+		}
+	}
+}
+
+/// A running instance of some `Subsystem`.
+struct SubsystemInstance<M: Debug, I> {
+	/// We talk to the `Overseer` over this channel.
+	rx: mpsc::Receiver<OverseerMessage<M, I>>,
+	/// The `Overseer` talks to use over this channel.
+	tx: mpsc::Sender<M>,
+}
+
+/// An `id` that is given to any `SubsystemInstance` for identification.
+#[derive(Clone, Copy, Debug, Eq, Hash, PartialEq)]
+pub struct SubsystemId(usize);
+
+/// A context type that is given to the `Subsystem` upon spawning
+/// that can be used by `Subsystem` to communicate with the outside world.
+pub struct SubsystemContext<M: Debug, I>{
+	rx: mpsc::Receiver<M>,
+	tx: mpsc::Sender<OverseerMessage<M, I>>,
+}
+
+impl<M: Debug, I> SubsystemContext<M, I> {
+	/// Try to asyncronously receive a message.
+	///
+	/// This has to be used with caution, if you loop over this without
+	/// using `pending!()` macro you will end up with a busy loop!
+	pub async fn try_recv(&mut self) -> Result<Option<M>, ()> {
+		match poll!(self.rx.next()) {
+			Poll::Ready(Some(msg)) => Ok(Some(msg)),
+			Poll::Ready(None) => Err(()),
+			Poll::Pending => Ok(None),
+		}
+	}
+
+	/// Receive a message.
+	pub async fn recv(&mut self) -> Result<M, SubsystemError> {
+		self.rx.next().await.ok_or(SubsystemError)
+	}
+
+	/// Send a message to whom it may concern.
+	///
+	/// The message will be broadcasted to all other `Subsystem`s that can
+	/// receive it.
+	pub async fn send_msg(&mut self, msg: M) {
```

Comment:
```suggestion
	pub async fn broadcast_msg(&mut self, msg: M) {
```

Wouldn't that name be more descriptive?